### DISCUSSION TITLE
Authenticated user dependency

### DISCUSSION THREAD

**Discussion by davidbrochart** [fastapi-users/fastapi-users#706] [Q&A]
I was wondering if the authenticated user could be used as a regular function call, by moving the callable from the router's dependencies:
```python
@app.get("/protected-route")
def protected_route(user: User = Depends(fastapi_users.current_user())):
    return f"Hello, {user.email}"
```
Inside the router's logic:
```python
@app.get("/protected-route")
async def protected_route():
    user: User = await fastapi_users.current_user()()
    return f"Hello, {user.email}"
```
My use-case is having a generic way of getting the current user, which is not known at router definition time (or can be changed later).

**Comment by frankie567** [comment]
The dependency is working inside the dependency injection of FastAPI, meaning it automatically injects the token from the authentication scheme. Thus, if you want to manually call the function, you have to pass the keyword argument corresponding to your authentication backend with the token value. Assuming you're using JWT, here is what it could look like:

```py
@app.get("/protected-route")
async def protected_route():
    user: User = await fastapi_users.current_user()(jwt="TOKEN")
    return f"Hello, {user.email}"
```

As a side note, I also strongly encourage you to assign the dependency in a variable instead of generating one over and over:

```py
get_current_user = fastapi_users.current_user()

@app.get("/protected-route")
async def protected_route():
    user: User = await get_current_user(jwt="TOKEN")
    return f"Hello, {user.email}"
```

It'll greatly help you if you need to override it in unit tests for example.
